prime-router/src/main/kotlin/serializers/Hl7Serializer.kt

```diff
@@ -202,65 +231,59 @@ class Hl7Serializer(val metadata: Metadata): Logging {
-                            mappedRows[it.name]?.add(decodeHl7DateTime(terser, it, warnings))
-                        }
-                        else -> {
-                            val terserSpec = when {
-                                it.hl7Field.startsWith("MSH") -> "/${it.hl7Field}"
-                                (it.hl7Field == "AOE") -> {
-                                    val question = it.hl7AOEQuestion!!
-                                    val countObservations = 10
-                                    // todo: map each AOE by the AOE question ID
-                                    for (c in 0 until countObservations) {
-                                        var spec = "/.OBSERVATION($c)/OBX-3-1"
-                                        val questionCode = try {
-                                            terser.get(spec)
-                                        } catch (e: HL7Exception) {
-                                            // todo: convert to result detail, maybe
-                                            errors.add("Exception for $spec: ${e.message}")
-                                            null
-                                        }
-                                        if (questionCode?.startsWith(question) == true) {
-                                            spec = "/.OBSERVATION($c)/OBX-5"
-                                            queryTerserForValue(terser, spec, it.name, mappedRows, errors, warnings)
-                                        }
-                                    }
-                                    "/.AOE"
-                                }
-                                else -> "/.${it.hl7Field}"
-                            }
-
-                            if (terserSpec != "/.AOE") {
-                                queryTerserForValue(terser, terserSpec, it.name, mappedRows, errors, warnings)
-                            } else {
-                                if (mappedRows[it.name]?.isEmpty() == true) mappedRows[it.name]?.add("")
-                            }
+            // Second, we process the mappers an only overwrite data values if the mapper returns a non-null string
+            schema.elements.forEach {
+                if (it.mapperRef != null) {
+                    // This gets the requiredvalue names, then gets the value from mappedRows that has the data
+                    val args = it.mapperArgs ?: emptyList()
+                    val valueNames = it.mapperRef.valueNames(it, args)
+                    val valuesForMapper = valueNames.mapNotNull { elementName ->
+                        val valueElement = schema.findElement(elementName)
+                        if(valueElement != null && mappedRows.containsKey(elementName) && !mappedRows[elementName].isNullOrEmpty()) {
+                            ElementAndValue(valueElement, mappedRows[elementName]!!.first())
+                        } else {
+                            null
```

Comment:
Hmm. The mapper algorithm should key on if a value is present and if a mapper is defined. The algorithm is: input_value ?: mapper value ?: default value.

Reply:
> Hmm. The mapper algorithm should key on if a value is present and if a mapper is defined. The algorithm is: input_value ?: mapper value ?: default value.

I can change that and look into making this common code.  However, I am wondering if we would want to have a way to post-process the data, like split it or validate it.